cl/beacon/validatorapi/post.go

```diff
@@ -46,14 +54,105 @@ func (v *ValidatorApiHandler) PostEthV1ValidatorSyncCommitteeSubscriptions(w htt
+	if len(req) == 0 {
+		return nil, nil
+	}
+
+	l := log.New()
+
+	/*
+		After beacon node receives this request,
+		search using discv5 for peers related to this subnet and replace current peers with those ones if necessary.
+		If validator is_aggregator, beacon node must:
+
+		 1. announce subnet topic subscription on gossipsub
+		 2. aggregate attestations received on that subnet
+	*/
+
+	// grab the head state to calculate committee counts with
+	s, err := v.privateGetStateFromStateId("head")
+	if err != nil {
+		return nil, err
+	}
+
+	digest, err := fork.ComputeForkDigest(v.BeaconChainCfg, v.GenesisCfg)
+	if err != nil {
+		l.Error("[Gossip] Failed to calculate fork choice", "err", err)
+	}
+
+	//calculate gossip_subnets
+	topics := make([]string, 0, len(req))
+	for _, x := range req {
+		// skip ones we are aggregator for
+		if !x.IsAggregator {
+			continue
+		}
+		committees_per_slot := s.CommitteeCount(uint64(x.Slot) / v.BeaconChainCfg.SlotsPerEpoch)
+		subnet := v.BeaconChainCfg.ComputeSubnetForAttestation(committees_per_slot, uint64(x.Slot), uint64(x.CommitteeIndex))
+		topicString := fmt.Sprintf("/eth2/%x/beacon_attestation_%d/%s", digest, subnet, "ssz_snappy")
+		topics = append(topics, topicString)
+	}
+
+	filterString := "/eth2/*/beacon_attestation_*/ssz_snappy"
+	subscriptionData := &sentinel.SubscriptionData{
+		Filter: &filterString,
+		Topics: topics, // TODO: this should be done in advance in the sentinel and recalculated every once in a while
+	}
+	subDuration := time.Duration(v.BeaconChainCfg.SecondsPerSlot*v.NetworkCfg.AttestationPropagationSlotRange*2) * time.Second
+	ctx, cn := context.WithTimeout(r.Context(), subDuration)
+	sub, err := v.Sentinel.SubscribeGossip(ctx, subscriptionData)
+	if err != nil {
+		cn()
+		return nil, err
+	}
+	go func() {
+		l.Info("[beacon api] started aggregating gossip", "subscription", subscriptionData)
+		defer func() {
+			l.Info("[beacon api] aggregation gossip subscriber closed")
+			cn()
+		}()
+		for {
+			dat, err := sub.Recv()
+			if err != nil {
+				return
+			}
+			split := strings.Split(strings.Trim(dat.Name, "/"), "/")
+			if len(split) < 4 {
+				continue
+			}
+			decimalString := strings.TrimPrefix(split[2], "beacon_attestation_")
+			// TODO: validate subnet topic int so we dont repeat work perhaps? not sure.
+			// possible optimization for cpu usage down the line
+			subnetTopicInt, err := strconv.ParseUint(decimalString, 10, 64)
+			if err != nil {
+				continue
+			}
+			l = l.New("subnetTopic", subnetTopicInt)
+			a := &solid.Attestation{}
+			err = a.DecodeSSZ(dat.Data, 0)
+			if err != nil {
+				l.Info("[beacon api] received invalid attestation", "error", err)
+				continue
+			}
+			//
+			err = v.Machine.ProcessAttestations(s, solid.NewDynamicListSSZFromList[*solid.Attestation]([]*solid.Attestation{a}, 1))
+			if err != nil {
+				l.Info("[beacon api] attestation failed validation", "error", err)
+				continue
+			}
+			v.state.AddAttestation(subnetTopicInt, a)
```

Comment:
why not add it to the operations pool? we have that.

Reply:
because i wanted to key them by the topic int as well.

Reply:
uhm sure, is it needed later on in the API though? i should check that